Port sumpy's cse implementation to sympy

isuruf opened the pull request:
This uses @mattwala's code in sumpy. I've also bundled an implementation of OrderedSet from http://code.activestate.com/recipes/576694/ which was suggested in python docs. OrderedSet code is licensed MIT. Is that okay to have without changing sympy's license?

Fixes #12411

isuruf committed:
Add sumpy's changes

isuruf committed:
Add OrderedSet

isuruf committed:
Use OrderedSet

isuruf committed:
Add recent sumpy changes

isuruf commented on sympy/simplify/cse_main.py:
```diff
@@ -136,6 +137,235 @@ def postprocess_for_cse(expr, optimizations):
+        self.func_to_argset[func_i].update(new_args)
+
+
+class Unevaluated(object):
```
This was initially added for symengine compatibility in sumpy, but this also has a performance benefit in that we can identify which expressions are needed to be evaluated at the end of the cse. (https://github.com/sympy/sympy/pull/13220/files#diff-44a512e7d608776ee035c0ab6a4baa75L411 was removed because of this)

ylemkimon commented on sympy/core/containers.py:
```diff
@@ -264,3 +266,66 @@ def __lt__(self, other):
+
+
+# OrderedSet code is based on http://code.activestate.com/recipes/576694/ by
+# Raymond Hettinger. Licensed under MIT License
```
If you've used MIT licensed code, you should include a copy of copyright notice, which looks like:
'''
Copyright <YEAR> <COPYRIGHT HOLDER>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
'''

isuruf committed:
Fix tests for cse implementation differences

isuruf committed:
Fix python 3.3 bug

isuruf commented:
Hmm, this PR uses `sets` in addition to `OrderedSet` and that introduces randomness which in turn affects the cse output. Is that okay?

isuruf committed:
Fix for test with random cse output

smichr commented:
> introduces randomness which in turn affects the cse output. Is that okay?

No. Wherever the items in the sets are iterated over, the set, S, should be replaced with ordered(S).

isuruf committed:
Use OrderedSet everywhere in cse_main

isuruf committed:
Revert "Fix for test with random cse output"

This reverts commit c74b4a49afd5244b87a42ade03b5b9ecb7eade0b.

isuruf committed:
Fix test

smichr commented on sympy/simplify/cse_main.py:
```diff
@@ -445,7 +445,7 @@ def _find_opts(expr):
 
     # split muls into commutative
-    comutative_muls = set()
+    comutative_muls = OrderedSet()
```
commutative -> commutative (I know that python doesn't care, but if we're going to spell out this much, let's go ahead and give it an M&M. :-))

smichr commented on sympy/simplify/cse_main.py:
```diff
@@ -491,10 +491,10 @@ def tree_cse(exprs, symbols, opt_subs=None, order='canonical', ignore=()):
-    seen_subexp = set()
-    excluded_symbols = set()
+    seen_subexp = OrderedSet()
+    excluded_symbols = OrderedSet()
```
Hey, what's happening down at line 527 where we have `list(map(...))` not being assigned to anything?

isuruf replied:
map is an iterator in python3 that's why there is a `list` call to make sure it's evaluated.

smichr commented on sympy/simplify/cse_main.py:
```diff
@@ -491,10 +491,10 @@ def tree_cse(exprs, symbols, opt_subs=None, order='canonical', ignore=()):
 
-    seen_subexp = set()
-    excluded_symbols = set()
+    seen_subexp = OrderedSet()
```
No need for OrderedSet here as far as I can see -- the sets are used for containment checks and are not iterated over. But double check.

isuruf commented on sympy/simplify/cse_main.py:
```diff
@@ -555,6 +744,9 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
     if postprocess is None:
-        return replacements, reduced_exprs
+        result = replacements, reduced_exprs
+        print(result[1].args[1].args)
```
Adding this print statement here, makes the test failing in the previous commit pass. Any ideas

isuruf replied:
It doesn't. Ignore my above comment

isuruf committed:
Fix solving odes

ODE solving used cse and to extract the Constant terms in the ODEs
and then replaced terms like C2/2 with C2. Issue was that C2 can
occur in multiple places and therefore replacing is not valid

isuruf committed:
typo

isuruf committed:
Remove unnecesary use of OrderedSet

smichr commented:
I see there was not much of a change in terms of expected results. What are the benefits of this new implementation?

asmeurer commented:
You shouldn't change the git author like that. That's effectively dishonest, because he never created that commit.

isuruf commented:
Okay. I'll fix that and add the license above.

asmeurer commented:
Also I suppose one could use OrderedDict as a proxy for OrderedSet, where each value is None. It might be faster since OrderedDict is written in C.

isuruf commented:
`OrderedDict` wouldn't have intersection, union defined.

isuruf committed:
Use OrderedDict for OrderedSet

isuruf committed:
Empty OrderedSet and repr

isuruf committed:
Add pop, intersection, difference and update

isuruf commented on sympy/core/containers.py:
```diff
@@ -264,3 +266,53 @@ def __lt__(self, other):
+        if iterable:
+            self.map = collections.OrderedDict((item, None) for item in iterable)
+        else:
+            self.map = collections.OrderedDict()
```
@asmeurer, thanks for the idea. This implementation is much simpler.

asmeurer commented:
The `odict_keys` objects support set operations (at least in Python 3), though it looks like they return unordered `sets`.  So maybe there is room for OrderedSet in the Python collections module. Originally, before Python had a `set` object people used this dict with None values trick (see https://www.youtube.com/watch?v=V5-JH23Vk0I).

isuruf commented:
@asmeurer, now that the OrderedSet code is not used, there's no license issue. Is there anything to be done regarding the license of sumpy code? You mentioned that there is no need to do anything at https://github.com/sympy/sympy/issues/12411#issuecomment-309848494. Just double checking.

asmeurer commented:
@mattwala agreed to relicense the code, so there are no issues.

mattwala commented on sympy/core/containers.py:
```diff
@@ -264,3 +266,53 @@ def __lt__(self, other):
+
+    def intersection(self, other):
+        result = []
+        for val in self:
```
Suggestion: The outer loop look should through the smaller of the two sets to reduce the number of lookups that are needed.

mattwala replied:
Never mind, that does not preserve the order.

smichr commented on sympy/simplify/cse_main.py:
```diff
@@ -136,6 +137,238 @@ def postprocess_for_cse(expr, optimizations):
+        """
+        return [self.value_number_to_value[argn] for argn in sorted(argset)]
+
+    def get_or_add_value_number(self, value):
```
rather than introducing a new name, can we just use the already known `setdefault` as is used for dictionaries?

```
>>> d = {1: 2}
>>> d.setdefault(1, 0)
2
>>> d.setdefault(3, 0)
0
>>> d
{1: 2, 3: 0}
```

smichr replied:
This would introduce variation wrt sumpy...which might get updated and then it might be harder to update this. So I'm OK with it as it is. Are you ready to commit this?

isuruf replied:
Yes, it'll be easier if the implementations were same.

smichr commented on sympy/simplify/cse_main.py:
```diff
@@ -136,6 +137,238 @@ def postprocess_for_cse(expr, optimizations):
+
+    def __init__(self, funcs):
+        # To minimize the number of symbolic comparisons, all function arguments
+        # get assigned a value number.
```
Although I tend to *think* that this matters, I thought that a hash code or something like that is used to tell if two items are the same or not. But a little timing indicates that it matters a LOT:

```
>>> timeit('1==3')
0.052123287547175275
>>> timeit('x==y','from sympy import S;from sympy.abc import x,y;x=S(1);y=S(2)')
1.5453416996564044
>>> timeit('x==y','from sympy.abc import x,y')
12.901627279220296
```

isuruf commented:
@moorepants, @bjodah would you mind running the benchmarks on this PR? Additionally, I can run the benchmarks if you can give me instructions on how to do so.

isuruf commented:
commit  | cse.TimeCSE.time_cse | large_exprs.TimeLargeExpressionOperations.time_cse
-----------|---------------------------------|-------------------------------------------------------------------------
This PR | 3.48 ms | 65.28 ms
master | 4.07 ms | 22.24 s
sympy-1.1.1 | 4.08 ms | 189.66 ms
sympy-1.0 | 4.05 ms | 195.33 ms

smichr commented:
OK, I don't see any reason to not move forward with this and nobody else has raised objections. This is in.

moorepants commented:
Oddly I'm seeing a speed regression from this on my benchmarks, from https://github.com/sympy/sympy/commit/0852d92d

http://www.moorepants.info/misc/sympy-asv/#cse.TimeCSE.time_cse?commits=76615a4e-fb9bf779&commits=0852d92d&idx=5

isuruf commented:
That's weird. I ran the same benchmark and there was a speedup. (I only checked the total time for that benchmark reported by asv, but it seems there is  a slow down for all parameters of the benchmark)

moorepants commented:
When I get back to my office Monday I'll run them on this machine again. I'm not sure if it matters, but I do have the cache size set quite high on this machine.

moorepants commented:
It looks like for python 2.7 this implementation is slower and for python 3.6 it is faster.

asmeurer commented:
Are you sure it's this PR. I know there have been general slowdowns in Python 2.7 since https://github.com/sympy/sympy/pull/12945

moorepants commented:
If you zoom in to the portion after Bjorn reverted that cse change that slowed things down, you'll see it: http://www.moorepants.info/misc/sympy-asv/#cse.TimeCSE.time_cse

You can flip the switches on the left to focus on a single py2.7 to py3.6 comparison. It points right to this commit.

moorepants commented:
And yes, #12945 has contributed to slowdowns in a number of the benchmarks.
